File: python/day4.py
import re

with open("inputs/4.txt") as file:
    data = file.read().strip()

passports = [passport.replace("\n", " ") for passport in data.split("\n\n")]
required = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]  # cid


def valid_1(passport: str) -> bool:
    return all(field in passport for field in required)


def valid_2(passport: str) -> bool:
    if not valid_1(passport):
        return False

    p = {field.split(":")[0]: field.split(":")[1] for field in passport.split(" ")}

    if not 1920 <= int(p["byr"]) <= 2002:
        return False

    if not 2010 <= int(p["iyr"]) <= 2020:
        return False

    if not 2020 <= int(p["eyr"]) <= 2030:
        return False

    height = int(p["hgt"][:-2])
    height_units = p["hgt"][-2:]
    if height_units == "cm":
        if not 150 <= height <= 193:
            return False
    elif height_units == "in":
        if not 59 <= height <= 76:
            return False
    else:
        return False

    if re.match(r"#[0-9a-f]{6}", p["hcl"]) is None:
        return False

    if p["ecl"] not in ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]:
        return False

    if len(p["pid"]) != 9 or not p["pid"].isnumeric():
        return False

    return True


print(sum(valid_1(passport) for passport in passports))
print(sum(valid_2(passport) for passport in passports))

File: rust/src/day4.rs
use std::collections::HashMap;
use regex::Regex;
use std::fs::read_to_string;

type Passport<'a> = HashMap<&'a str, &'a str>;

lazy_static! {
    static ref RE_PASSPORT :Regex =Regex::new(r"(?P<field>[a-z]+):(?P<value>\S+)").unwrap();
}

fn parse(input: &str) -> Vec<Passport> {
    input.split("\n\n")
        .map(|passport| RE_PASSPORT.captures_iter(passport.as_ref())
            .map(|captures| (
                captures.name("field").unwrap().as_str(),
                captures.name("value").unwrap().as_str()
            )).collect())
        .collect()
}

pub fn solve() {
    let input = read_to_string("../inputs/4.txt").unwrap();
    let passports = parse(&input);
    print!("Day 4 part 1: {}\n", part_1(&passports));
    print!("Day 4 part 2: {}\n", part_2(&passports));
}

fn valid1(passport: &&Passport) -> bool {
    ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]
        .iter()
        .all(|key| (**passport).contains_key(key))
}

fn valid2(passport: &&Passport) -> bool {
    ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]
        .iter()
        .all(|key| (**passport).contains_key(key))
}

fn part_1(passports: &Vec<Passport>) -> usize {
    passports.iter().filter(valid1).count()
}

fn part_2(passports: &Vec<Passport>) -> usize {
    passports.iter().filter(valid2).count()
}

#[cfg(test)]
mod tests {
    use std::fs::read_to_string;
    use super::parse;
    use super::part_1;
    use super::part_2;

    #[test]
    fn test_parser() {
        let data = "ecl:gry pid:860033327\n\neyr:2020";
        let passports = parse(&data);
        assert_eq!(passports[0]["ecl"], "gry");
        assert_eq!(passports[0]["pid"], "860033327");
        assert_eq!(passports[1]["eyr"], "2020");
    }

    #[test]
    fn test_part1() {
        let data = "ecl:gry pid:860033327 eyr:2020 hcl:#fffffd\nbyr:1937 iyr:2017 cid:147 hgt:183cm\n\niyr:2013 ecl:amb cid:350 eyr:2023 pid:028048884\nhcl:#cfa07d byr:1929\n\nhcl:#ae17e1 iyr:2013\neyr:2024\necl:brn pid:760753108 byr:1931\nhgt:179cm\n\nhcl:#cfa07d eyr:2025 pid:166559648\niyr:2011 ecl:brn hgt:59in";
        let passports = parse(&data);
        let result = part_1(&passports);
        assert_eq!(result, 2);
    }

    #[test]
    fn test_part2() {
        let data = read_to_string("../inputs/1.txt").unwrap();
        let passports = parse(&data);
    }
}
